openedx/core/djangoapps/waffle_utils/tests/test_init.py

```diff
@@ -40,39 +43,98 @@ def setUp(self):
+        (True, WaffleFlagOrgOverrideModel.ALL_CHOICES.unset, True),
+        (False, WaffleFlagOrgOverrideModel.ALL_CHOICES.on, True),
+        (True, WaffleFlagOrgOverrideModel.ALL_CHOICES.on, True),
+        (False, WaffleFlagOrgOverrideModel.ALL_CHOICES.off, False),
+        (True, WaffleFlagOrgOverrideModel.ALL_CHOICES.off, False),
+    )
+    @ddt.unpack
+    def test_matching_org_override_waffle_flag(self, waffle_enabled, org_override_choice, is_enabled):
+        """
+        Tests various combinations of a flag being set in waffle and overridden for an org
+        which is the org which authored/owns the course.
+        Since the org-level override has the same org as the course being checked, the org-level
+        override's on/off/unset state determines whether is CourseWaffleFlag is active or not.
+
+        on    = active
+        off   = inactive
+        unset = mirror the base waffle flag's activity
+        """
+        WaffleFlagOrgOverrideModel.objects.create(
+            waffle_flag=self.NAMESPACED_FLAG_NAME,
+            org=self.TEST_ORG,
+            override_choice=org_override_choice,
+            note='',
+            enabled=True
+        )
+        # Both course keys should match the org-level override.
+        with override_flag(self.NAMESPACED_FLAG_NAME, active=waffle_enabled):
+            assert self.TEST_COURSE_FLAG.is_enabled(self.TEST_COURSE_KEY) == is_enabled
+            assert self.TEST_COURSE_FLAG.is_enabled(self.TEST_COURSE_2_KEY) == is_enabled
+
+    @ddt.data(
+        (False, WaffleFlagOrgOverrideModel.ALL_CHOICES.unset, False),
+        (True, WaffleFlagOrgOverrideModel.ALL_CHOICES.unset, True),
+        (False, WaffleFlagOrgOverrideModel.ALL_CHOICES.on, False),
+        (True, WaffleFlagOrgOverrideModel.ALL_CHOICES.on, True),
+        (False, WaffleFlagOrgOverrideModel.ALL_CHOICES.off, False),
+        (True, WaffleFlagOrgOverrideModel.ALL_CHOICES.off, True),
+    )
+    @ddt.unpack
+    def test_not_matching_org_override_waffle_flag(self, waffle_enabled, org_override_choice, is_enabled):
+        """
+        Tests various combinations of a flag being set in waffle and overridden for an org
+        which is *not* the target course's org.
+        Since the org-level override isn't relevant to the course being checked, whether the
+        waffle flag is active/inactive determines whether the CourseWaffleFlag is active or not.
+
+        on/off/unset = mirror the base waffle flag's activity
+        """
+        WaffleFlagOrgOverrideModel.objects.create(
+            waffle_flag=self.NAMESPACED_FLAG_NAME,
+            org=self.TEST_ORG,
+            override_choice=org_override_choice,
+            note='',
+            enabled=True
+        )
+        # Org doesn't match the course key, so should never be enabled.
+        with override_flag(self.NAMESPACED_FLAG_NAME, active=waffle_enabled):
+            assert self.TEST_COURSE_FLAG.is_enabled(self.TEST_COURSE_3_KEY) == is_enabled
```

Comment:
Do we need more tests on the interaction between Org and Course level flags, like:

```python
(WaffleFlagOrgOverrideModel.ALL_CHOICES.unset, WaffleFlagCourseOverrideModel.ALL_CHOICES.unset, False)
(WaffleFlagOrgOverrideModel.ALL_CHOICES.unset, WaffleFlagCourseOverrideModel.ALL_CHOICES.on, True)
# etc.
```

Reply:
Yes - agreed. I'll add these tests with some explanation of their expected results.

Reply:
Added now.